Update Fetching Sdk Objects

nahara7 opened the pull request:
Addresses: #94 

Allows Submissions and Replies to be retrieved by only their `uuid` and `source.uuid`
### Testing

Tested against the client and development server using the following branch: https://github.com/nahara7/securedrop-client/tree/nahara/test-sdk and modifying the download call[ here](https://github.com/nahara7/securedrop-client/blob/c6959f36252e85bd44b14a8d6f15a375b6b259f2/securedrop_client/api_jobs/downloads.py#L257.) and[ here](https://github.com/nahara7/securedrop-client/blob/c6959f36252e85bd44b14a8d6f15a375b6b259f2/securedrop_client/api_jobs/downloads.py#L257.):

creviera reviewed (dismissed):
Looks good to me and works as described. Thanks @nahara7 :tada: 

## Test Plan
- [x] Test for regressions when sending and receiving replies and submissions against the client
- [x] Test for regressions when sending and receiving replies and submissions against the client with these changes applied: https://github.com/nahara7/securedrop-client/commit/c6959f36252e85bd44b14a8d6f15a375b6b259f2

creviera commented:
Here's the test plan I'm using for this PR, and my results so far (I've only been able to functional test and ran out of time to code review today):

**Test against current client**
0. [Run a Securedrop dev server](https://github.com/freedomofpress/securedrop#developer-quickstart)
1. Checkout and pull the latest changes from the `main` branch in the `securedrop-client` repo
2. `make venv && source .venv/bin/activate`
3. `LOGLEVEL=debug ./run.sh --sdc-home ~/.securedrop-client` and tail the logs in `~/.securedrop_client/logs/client.log`
4. Log into the client as `journalist`
5. From the Source Interface (http://127.0.0.1:8080/) send a file attachment and message
6. From the Journalist Interface (http://127.0.0.1:8081/), log in and send a reply to that source
- [x] Select that source in the client and confirm that you see the decrypted message and reply
- [x] Download the file and confirm you see the filename with the options to print and export (you won't be able to test these operations outside of QubesOS)

**Test new SDK API**
0. [Run a Securedrop dev server](https://github.com/freedomofpress/securedrop#developer-quickstart)
1. Checkout and pull the latest changes from the `nahara/test-sdk` branch in the forked `securedrop-client` repo
2. `make venv && source .venv/bin/activate`
3. `pip uninstall securedrop-sdk && pip install git+https://github.com/nahara7/securedrop-sdk@nahara/sdk-objects#egg=securedrop-sdk`
4. `LOGLEVEL=debug ./run.sh --sdc-home ~/.securedrop-client` and tail the logs in `~/.securedrop_client/logs/client.log`
5. Log into the client as `journalist`
6. From the Source Interface (http://127.0.0.1:8080/) send a file attachment and message
7. From the Journalist Interface (http://127.0.0.1:8081/), log in and send a reply to that source
- :x: Select that source in the client and confirm that you see the decrypted message and reply
  > The reply does not decrypt
- [ ] Download the file and confirm you see the filename with the options to print and export (you won't be able to test these operations outside of QubesOS)

creviera commented:
> 8. Select that source in the client, and
> :x: Confirm that you see the decrypted message and reply
> (The reply does not decrypt.)

Is this what you're experiencing as well, @nahara7?

creviera commented:
As discussed in gitter, adding log lines around https://github.com/freedomofpress/securedrop-sdk/blob/main/sdclientapi/__init__.py#L803 will help us understand more about the underlying issue.

nahara7 commented:
> > 8. Select that source in the client, and
> >    ❌ Confirm that you see the decrypted message and reply
> >    (The reply does not decrypt.)
> 
> Is this what you're experiencing as well, @nahara7?

I see the messages and replies as decrypted when I follow this testing plan.

nahara7 commented:
Per our discussion today @creviera I have a better understanding of the testing plan. I can confirm I have the same issues where replies stay encrypted in step 8. I am also working on debugging the following error, which I believe is related to my windows environment. There is aslo a `DownloadDecryptionFailure`.  To continue I'll open a debugging branch.
```
    self.socket.send(msg)
OSError: [Errno 9] Bad file descriptor
...
During handling of the above exception, another exception occurred:    

Traceback (most recent call last):
  File "/usr/lib/python3.7/logging/handlers.py", line 855, in _connect_unixsocket
    self.socket.connect(address)
FileNotFoundError: [Errno 2] No such file or directory

During handling of the above exception, another exception occurred:    

Traceback (most recent call last):
  File "/usr/lib/python3.7/logging/handlers.py", line 937, in emit     
    self._connect_unixsocket(self.address)
  File "/usr/lib/python3.7/logging/handlers.py", line 866, in _connect_unixsocket
    self.socket.connect(address)
FileNotFoundError: [Errno 2] No such file or directory
Call stack:
  File "/mnt/c/Users/Nahara/OneDrive/Documents/GitHub/securedrop-client/securedrop_client/queue.py", line 180, in process
    logger.error("{}: {}".format(type(e).__name__, e))
Message: 'DownloadDecryptionException: Failed to download 52a52c73-9701-4a56-b13f-64cb12095fee'
Arguments: ()
--- Logging error ---
Traceback (most recent call last):
  File "/mnt/c/Users/Nahara/OneDrive/Documents/GitHub/securedrop-client/securedrop_client/api_jobs/downloads.py", line 148, in _download      
    etag, download_path = self.call_download_api(api, db_object)       
  File "/mnt/c/Users/Nahara/OneDrive/Documents/GitHub/securedrop-client/securedrop_client/api_jobs/downloads.py", line 263, in call_download_api
    return api.download_reply(sdk_object)
  File "/mnt/c/Users/Nahara/OneDrive/Documents/Github/securedrop-client/.venv/src/securedrop-sdk/sdclientapi/__init__.py", line 842, in download_reply
    with open(filepath, "wb") as fobj:
FileNotFoundError: [Errno 2] No such file or directory: ''

The above exception was the direct cause of the following exception:   

Traceback (most recent call last):
  File "/mnt/c/Users/Nahara/OneDrive/Documents/GitHub/securedrop-client/securedrop_client/queue.py", line 166, in process
    self.current_job._do_call_api(self.api_client, session)
  File "/mnt/c/Users/Nahara/OneDrive/Documents/GitHub/securedrop-client/securedrop_client/api_jobs/base.py", line 73, in _do_call_api
    result = self.call_api(api_client, session)
  File "/mnt/c/Users/Nahara/OneDrive/Documents/GitHub/securedrop-client/securedrop_client/api_jobs/downloads.py", line 136, in call_api       
    destination = self._download(api_client, db_object, session)       
  File "/mnt/c/Users/Nahara/OneDrive/Documents/GitHub/securedrop-client/securedrop_client/api_jobs/downloads.py", line 175, in _download      
    ) from e
securedrop_client.api_jobs.downloads.DownloadDecryptionException: Failed to download 52a52c73-9701-4a56-b13f-64cb12095fee

During handling of the above exception, another exception occurred:    

Traceback (most recent call last):
  File "/usr/lib/python3.7/logging/handlers.py", line 934, in emit     
    self.socket.send(msg)
OSError: [Errno 9] Bad file descriptor

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.7/logging/handlers.py", line 855, in _connect_unixsocket
    self.socket.connect(address)
FileNotFoundError: [Errno 2] No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.7/logging/handlers.py", line 937, in emit
    self._connect_unixsocket(self.address)
  File "/usr/lib/python3.7/logging/handlers.py", line 866, in _connect_unixsocket
    self.socket.connect(address)
FileNotFoundError: [Errno 2] No such file or directory
Call stack:
  File "/mnt/c/Users/Nahara/OneDrive/Documents/GitHub/securedrop-client/securedrop_client/queue.py", line 181, in process      
    logger.debug("Skipping job")
Message: 'Skipping job'```

creviera commented:
Thanks for sharing your logs @nahara7! It looks like the `DownloadDecryptionFailure` is the result of the `FileNotFoundError` in the SDK bubbling up. The SDK can't find the file it needs here: https://github.com/freedomofpress/securedrop-sdk/blob/f43951c8dbc08067a1bedf43258ef5d08361b781/sdclientapi/__init__.py#L842 But I'm noticing this is a bug that would only happen in a development environment because the code that's failing is inside a conditional that checks if we're not using a proxy (in a production environment, one has to use the proxy).

Since we're trying to remove this filename parameter that appears only to be used in a dev environment, we'll have to look at how to get the filename for the reply from the response instead. Just as a quick test, I updated the sdk in the client's virtual env: `.venv/liv/python3.7/site-packages/sdclientapi/__init__.py` to (1) inspect the returned `headers` parameter by throwing in a print statement [before this line of code](https://github.com/freedomofpress/securedrop-sdk/blob/f43951c8dbc08067a1bedf43258ef5d08361b781/sdclientapi/__init__.py#L841), and (2) updated `reply.filename` to `headers['Content-Disposition'].split('attachment; filename=')[1]` which got this working and decrypting replies. Seems like a bit of a hack, but better than requiring a dev-use-only filename parameter. Are you able to confirm this fix? If it works, let's consider cleaning it up and using it, and add a prod test scenario to your PR's test plan (just in case).

nahara7 commented:
> Thanks for sharing your logs @nahara7! It looks like the `DownloadDecryptionFailure` is the result of the `FileNotFoundError` in the SDK bubbling up. The SDK can't find the file it needs here:
> 
> https://github.com/freedomofpress/securedrop-sdk/blob/f43951c8dbc08067a1bedf43258ef5d08361b781/sdclientapi/__init__.py#L842
> 
> But I'm noticing this is a bug that would only happen in a development environment because the code that's failing is inside a conditional that checks if we're not using a proxy (in a production environment, one has to use the proxy).
> Since we're trying to remove this unnecessary filename parameter (and rely on source_uuid + the uuid for the reply instead) that appears only to be used in a dev environment, we'll have to look at how to get the filename for the reply from the response instead. Just as a quick test, I updated the sdk in the client's virtual env: `.venv/liv/python3.7/site-packages/sdclientapi/__init__.py` to (1) inspect the returned `headers` parameter by throwing in a print statement [before this line of code](https://github.com/freedomofpress/securedrop-sdk/blob/f43951c8dbc08067a1bedf43258ef5d08361b781/sdclientapi/__init__.py#L841), and (2) updated `reply.filename` to `headers['Content-Disposition'].split('attachment; filename=')[1]` which got this working and decrypting replies. Seems like a bit of a hack, but better than requiring a dev-use-only filename requirement when the client wants to download a reply from the server. Are you able to confirm this fix? If it works, let's consider cleaning it up and using it, and add a prod test scenario to your PR's test plan (just in case).

Yes this works for me ! Awesome ! I'm updating the branch. How will a prod test scenario plan differ from the plan you outlined above ?

nahara7 commented:
> Thanks for sharing your logs @nahara7! It looks like the `DownloadDecryptionFailure` is the result of the `FileNotFoundError` in the SDK bubbling up. The SDK can't find the file it needs here:
> 
> https://github.com/freedomofpress/securedrop-sdk/blob/f43951c8dbc08067a1bedf43258ef5d08361b781/sdclientapi/__init__.py#L842
> 
> But I'm noticing this is a bug that would only happen in a development environment because the code that's failing is inside a conditional that checks if we're not using a proxy (in a production environment, one has to use the proxy).
> Since we're trying to remove this unnecessary filename parameter (and rely on source_uuid + the uuid for the reply instead) that appears only to be used in a dev environment, we'll have to look at how to get the filename for the reply from the response instead. Just as a quick test, I updated the sdk in the client's virtual env: `.venv/liv/python3.7/site-packages/sdclientapi/__init__.py` to (1) inspect the returned `headers` parameter by throwing in a print statement [before this line of code](https://github.com/freedomofpress/securedrop-sdk/blob/f43951c8dbc08067a1bedf43258ef5d08361b781/sdclientapi/__init__.py#L841), and (2) updated `reply.filename` to `headers['Content-Disposition'].split('attachment; filename=')[1]` which got this working and decrypting replies. Seems like a bit of a hack, but better than requiring a dev-use-only filename requirement when the client wants to download a reply from the server. Are you able to confirm this fix? If it works, let's consider cleaning it up and using it, and add a prod test scenario to your PR's test plan (just in case).

I followed the testing plan again, this time from the client-branch, using the SDK python egg. 

The replies are decrypted after adding the hack you outlined, though the logging errors do persist. I think this is unrelated to the changes and instead to my environment.

``` File "/usr/lib/python3.7/logging/handlers.py", line 937, in emit  
    self._connect_unixsocket(self.address)
  File "/usr/lib/python3.7/logging/handlers.py", line 866, in _connect_unixsocket
    self.socket.connect(address)
FileNotFoundError: [Errno 2] No such file or directory
Call stack:
  File "/mnt/c/Users/Nahara/OneDrive/Documents/GitHub/securedrop-client/securedrop_client/sync.py", line 137, in sync
    self.job._do_call_api(self.api_client, session)
  File "/mnt/c/Users/Nahara/OneDrive/Documents/GitHub/securedrop-client/securedrop_client/api_jobs/base.py", line 73, in _do_call_api   
    result = self.call_api(api_client, session)
  File "/mnt/c/Users/Nahara/OneDrive/Documents/GitHub/securedrop-client/securedrop_client/api_jobs/sync.py", line 46, in call_api       
    sources, submissions, replies = get_remote_data(api_client)     
  File "/mnt/c/Users/Nahara/OneDrive/Documents/GitHub/securedrop-client/securedrop_client/storage.py", line 121, in get_remote_data     
    logger.info("Fetched {} remote replies.".format(len(remote_replies)))
Message: 'Fetched 7 remote replies.'
Arguments: ()
--- Logging error ---
Traceback (most recent call last):
  File "/usr/lib/python3.7/logging/handlers.py", line 934, in emit  
    self.socket.send(msg)
OSError: [Errno 9] Bad file descriptor

During handling of the above exception, another exception occurred: 

Traceback (most recent call last):
  File "/usr/lib/python3.7/logging/handlers.py", line 855, in _connect_unixsocket
    self.socket.connect(address)
FileNotFoundError: [Errno 2] No such file or directory

During handling of the above exception, another exception occurred: 

Traceback (most recent call last):
  File "/usr/lib/python3.7/logging/handlers.py", line 937, in emit  
    self._connect_unixsocket(self.address)
  File "/usr/lib/python3.7/logging/handlers.py", line 866, in _connect_unixsocket
    self.socket.connect(address)
FileNotFoundError: [Errno 2] No such file or directory
Call stack:
  File "/mnt/c/Users/Nahara/OneDrive/Documents/GitHub/securedrop-client/securedrop_client/sync.py", line 137, in sync
    self.job._do_call_api(self.api_client, session)
  File "/mnt/c/Users/Nahara/OneDrive/Documents/GitHub/securedrop-client/securedrop_client/api_jobs/base.py", line 73, in _do_call_api   
    result = self.call_api(api_client, session)
  File "/mnt/c/Users/Nahara/OneDrive/Documents/GitHub/securedrop-client/securedrop_client/api_jobs/sync.py", line 47, in call_api       
    update_local_storage(session, sources, submissions, replies, self.data_dir)
  File "/mnt/c/Users/Nahara/OneDrive/Documents/GitHub/securedrop-client/securedrop_client/storage.py", line 184, in update_local_storage    remote_sources, get_local_sources(session), skip_conversation_uuids, session, data_dir
  File "/usr/lib/python3.7/contextlib.py", line 119, in __exit__    
    next(self.gen)
  File "/mnt/c/Users/Nahara/OneDrive/Documents/GitHub/securedrop-client/securedrop_client/utils.py", line 204, in chronometer
    logger.info(f"{description} duration: {elapsed:.4f}s")
Message: 'update_sources duration: 0.0041s'
Arguments: ()
--- Logging error ---
Traceback (most recent call last):
  File "/usr/lib/python3.7/logging/handlers.py", line 934, in emit  
    self.socket.send(msg)
OSError: [Errno 9] Bad file descriptor```

creviera commented:
@nahara7 I'll be able to take a closer look early next week, but if it's not too much trouble it might be helpful to set up a Debian or Ubuntu development environment. And since you've already figured out how to get this working on Windows, hopefully it'll be easier the second time around (if not, there are folks in https://gitter.im/freedomofpress/securedrop that will be able to help).

nahara7 commented:
> @nahara7 I'll be able to take a closer look early next week, but if it's not too much trouble it might be helpful to set up a Debian or Ubuntu development environment. And since you've already figured out how to get this working on Windows, hopefully it'll be easier the second time around (if not, there are folks in https://gitter.im/freedomofpress/securedrop that will be able to help).

 For the above trace it looks like my syslog service wasn't started.  Looks good now !
